wicker/core/datasets.py

```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
+        self._path_factory = path_factory
+        self._schema: Optional[DatasetSchema] = None  # Set by lazy initialization
+        self._storage = storage
         self._treat_objects_as_bytes = treat_objects_as_bytes
-        self._schema: Optional[DatasetSchema] = None
-        self._arrow_table: Optional[pyarrow.Table] = None
 
-        self._local_cache_path_prefix = local_cache_path_prefix
-        self._filelock_timeout_seconds = filelock_timeout_seconds
-        self._storage = storage if storage is not None else S3DataStorage()
-        self._s3_path_factory = s3_path_factory if s3_path_factory is not None else S3PathFactory()
-        self._column_bytes_file_cache = ColumnBytesFileCache(
-            local_cache_path_prefix=local_cache_path_prefix,
-            filelock_timeout_seconds=filelock_timeout_seconds,
-            path_factory=self._s3_path_factory,
-            storage=self._storage,
-            dataset_name=dataset_name,
-        )
-        self._pa_filesystem = (
-            pafs.S3FileSystem(region=get_config().aws_s3_config.region) if pa_filesystem is None else pa_filesystem
+        # Create the column bytes reader that depends on the other init parameter values.
+        self._column_bytes_file_reader = self._build_column_bytes_reader()
+
+    def __getitem__(self, idx: int) -> Dict[str, Any]:
+        """Get data item at index within arrow table.
+
+        Pulls from either cache or data store the item in the dataset at index specified.
+
+        Args:
+            idx (int): idx in arrow table to grab data.
+
+        Returns:
+            Dict[str, Any]: Row of data defined through schema object.
+        """
+        tbl = self.arrow_table()
+        columns = self._columns_to_load if self._columns_to_load is not None else tbl.column_names
+        row = {col: tbl[col][idx].as_py() for col in columns}
+        schema = self.schema()
+        return dataloading.load_example(
+            self._column_bytes_file_reader.resolve_pointers(row, schema),
+            schema,
         )
 
-        self._dataset_id = DatasetID(name=dataset_name, version=dataset_version)
-        self._partition = DatasetPartition(dataset_id=self._dataset_id, partition=dataset_partition_name)
-        self._dataset_definition = DatasetDefinition(
-            self._dataset_id,
-            schema=self.schema(),
+    def __len__(self) -> int:
+        """Get length of arrow table inferring it is the same as the dataset.
+
+        Returns:
+            int: length of arrow table.
+        """
+        return len(self.arrow_table())
+
+    def _build_column_bytes_reader(self) -> ColumnBytesFileReader:
+        """Builder function to return a column bytes file reader."""
+        if self._local_cache_path_prefix:
+            # If caching column files from storage into the local filesystem, return a cache-enabled reader.
+            return ColumnBytesFileCache(
```

Comment:
I would like to add a columfileReader para to the __init__ of the basedataset and push down this cache vs no-cache builder to the subclass. 
in the s3 dataset build the ColumnBytesFileCache add inject it the BaseDataset
in the filesystem dataset build either ColumnBytesFileCache or ColumnBytesFileReader based on the FuseFilesystem cache cap. By this mean, I would high suggest a get_dataset factory or builder method. For the new introduced FileSystemDataset, we should hide the detail to improve later. Otherwise, we just mix the Interface and Implementation together like the original Wicker S3Dataset, by that way. No point for an abstract dataset

Reply:
This make sense; I added the column bytes file reader parameter to the `BaseDataset.__init__` function and pushed down its construction to the derived classes as suggested above. The `S3Dataset` class will always build a a `ColumnBytesFileCache` instance. I think it's fine for now if the `FileSystemDataset` always builds a `ColumnBytesFileReader` instance.